### PR TITLE
feat: Pattern matching on arrays

### LINKED ISSUE #212: Lang: Pattern match on Arrays
Today, I realized that we didn't have an actual issue for implementing pattern matching on Arrays (just #189 that mentions it).

This would make a lot of code that works with arrays cleaner.

### PR DESCRIPTION
**Pull request by ospencer**
This PR adds the ability to pattern match on Arrays.

Closes #212.

### CODE REVIEW

File: compiler/src/middle_end/matchcomp.re [hunk 1]
```diff
@@ -485,6 +555,30 @@ let matrix_head_constructors = mtx => {
...
+  | TPatArray(ps) when !List.mem(List.length(ps), acc) => [
+      List.length(ps),
+      ...acc,
+    ]
+  | TPatOr(p1, p2) =>
+    pattern_head_arities_aux(p1, pattern_head_arities_aux(p2, acc))
+  | _ => acc
+  };
+
+let pattern_head_arities = patt => pattern_head_arities_aux(patt, []);
```
Comment (phated) [inline compiler/src/middle_end/matchcomp.re:569]: What's the difference between an `_aux` recursive method outside the func and the `let rec help` method like below?

File: compiler/src/middle_end/matchcomp.re [hunk 2]
```diff
@@ -627,6 +746,34 @@ let rec compile_matrix = mtx =>
...
+        (final_tree, [(arity, final_tree), ...switch_branches]);
+      };
+
+      switch (arities) {
+      | [] =>
+        failwith(
+          "Internal error: compile_matrix: non-wildcard column returned no arities",
+        )
+      | _ =>
+        let (result, switch_branches) =
```
Comment (phated) [inline compiler/src/middle_end/matchcomp.re:764]: Am I blind or is this `result` never used?

**Review by phated — APPROVED** [review]
👏 🎊 You know how excited I am to have this!! Amazing work!

File: compiler/src/middle_end/matchcomp.re (hunk 1 above)
Comment (ospencer) [reply compiler/src/middle_end/matchcomp.re:569]: None, really. I just followed the pattern of `pattern_head_constructors`.

File: compiler/src/middle_end/matchcomp.re (hunk 2 above)
Comment (ospencer) [reply compiler/src/middle_end/matchcomp.re:764]: It is not. I'll get that updated.

**Pull request merged**